rasa/core/actions/action.py

```diff
@@ -147,41 +125,53 @@ def is_retrieval_action(action_name: Text, retrieval_intents: List[Text]) -> boo
     )
 
 
-def action_from_name(
-    name: Text, domain: Domain, action_endpoint: Optional[EndpointConfig]
+def action_for_name_or_text(
+    action_name_or_text: Text, domain: Domain, action_endpoint: Optional[EndpointConfig]
 ) -> "Action":
     """Retrieves an action by its name.
 
     Args:
-        name: The name of the action.
+        action_name_or_text: The name of the action.
         domain: The current model domain.
         action_endpoint: The endpoint to execute custom actions.
 
+    Raises:
+        ActionNotFoundException: If action not in current domain.
+
     Returns:
         The instantiated action.
     """
+    if action_name_or_text not in domain.action_names:
+        domain.raise_action_not_found_exception(action_name_or_text)
+
     defaults = {a.name(): a for a in default_actions(action_endpoint)}
 
-    if name in defaults and name not in domain.user_actions_and_forms:
-        return defaults[name]
+    if (
+        action_name_or_text in defaults
+        and action_name_or_text not in domain.user_actions_and_forms
+    ):
+        return defaults[action_name_or_text]
 
-    if name.startswith(UTTER_PREFIX) and is_retrieval_action(
-        name, domain.retrieval_intents
+    if action_name_or_text.startswith(UTTER_PREFIX) and is_retrieval_action(
+        action_name_or_text, domain.retrieval_intents
     ):
-        return ActionRetrieveResponse(name)
+        return ActionRetrieveResponse(action_name_or_text)
 
-    if name.startswith(UTTER_PREFIX):
-        return ActionUtterTemplate(name)
+    if action_name_or_text in domain.action_texts:
```

Comment:
also added a test for this to ensure nobody thinks about switching these :-D

Reply:
but I'd move this `if` to line 156, do you think this check will hinder the performance significantly? The win is that it'll be less error prone

Reply:
tbh, I don't think the performance here matters too much. That should literally be a few milliseconds while e.g. some ML policies perform duplicate steps which are way more cost intense. Or we also fetch the tracker from the database more often than we need too which is in the region of ~100ms

Reply:
(recently started exploring prediction performance in https://github.com/RasaHQ/rasa/pull/7205)

Reply:
so why didn't you move it?

Reply:
it should be the second check before `utter_` prefix check

Reply:
@wochinge do we have a check that `action_text` is not exactly the same as some action name, then I believe will have a problem that `domain.action_name_or_text` contain non unique items

Reply:
Yes, `Domain` will check it 👍  Is your message missing a `not have a problem` or is it on purpose? You mean in case a message is coming up in several stories? We de-duplicate them before ([here](https://github.com/RasaHQ/rasa/blob/49bac016e71c8a053fceeb772f0d7f8764888d07/rasa/shared/importers/importer.py#L449))

Reply:
without the check there would be a problem, but if we have the check it should be good